File: antia_numerical_methods/c/xmp/drvt.c
/*	Differentiation using h ---> 0  extrapolation */

#include <stdio.h>
#include <math.h>
double drvt(double a, int id, double hh0, double reps, double aeps,
	double (*f)(double ), int *ier);

double fun(double x);

main()
{
	int i,i1,j, id,  ier;
	double xb, h, reps, aeps, df, fb[20];

/*	Example 5.2 : Derivatives of Exp(x) */

	aeps=1.e-6; reps=1.e-5;
	for(i1=0; i1<99; ++i1) {
		printf("type  id = order of derivative, x, h = initial spacing\n");
		printf("                              (quits when id<=0)\n");
		scanf("%d %le %le", &id, &xb, &h);
		if(id<=0) return 0;

		df=drvt(xb, id, h, reps, aeps, fun, &ier);
		printf(" ier = %d   id = %d   x = %e  df = %e \n", ier,id,xb,df);
		printf("  initial spacing = %e    exact derivative = %e\n",h,exp(xb));
	}
	return;
}

/*  The given function  */

double fun(double x)

{
	return exp(x);
}

 


/*	To calculate derivative of a function

	A : (input) Value of x at which the derivative needs to be calculated
	ID : (input) Order of derivative required, ID may be 1,2,3 or 4
	HH0 : (input) Initial value of step length to be used.
	REPS : (input) Required relative accuracy
	AEPS : (input) Required absolute accuracy.
		Calculations will continue till estimated error is less
		than max(AEPS, REPS*fabs(DRVT))
	F : (input) The name of function routine to calculate the given function
		Function F(X) must be supplied by the user.
	IER : (output) Error parameter, IER=0 for successful execution
		IER=28 implies required accuracy is not achieved
		IER=29 implies roundoff error appears to be dominating
		IER=208 implies ID<1 or ID>4, no calculations are done.
		
	DRVT is the calculated derivative at x=A

	Function F(X) must be supplied by the user

	Required functions : F
*/

#include <math.h>

double drvt(double a, int id, double hh0, double reps, double aeps,
	double (*f)(double ), int *ier)

{
	int i,j,ju, nmax=15, ncmax=6;
	double df, e1, err, err1, hh, h[15], t[15][6], hfac=1.5;

/*	Set spacing to the specified initial value */

	hh=hh0; err=0.0; *ier=0;

	for(i=0; i<nmax; ++i) {
		h[i]=hh;

		if(id == 1) t[i][0] = (f(a+hh)-f(a-hh))/(2.*hh);
		else if(id == 2) t[i][0] = (f(a-hh)-2.*f(a)+f(a+hh))/(hh*hh);
		else if(id == 3) t[i][0] = (f(a+2*hh)-f(a-2*hh)-2.*(f(a+hh)-f(a-hh)))/(2.*hh*hh*hh);
		else if(id == 4)
			t[i][0] = (f(a+2*hh)+f(a-2*hh)-4.*(f(a+hh)+f(a-hh))+6*f(a))/(hh*hh*hh*hh);
		else {*ier=208; return 0.0;}

/*	Reduce spacing by a factor of hfac */

		hh=hh/hfac;
		if(i>0) {
/*	Calculate at most ncmax columns of the T-table */
			ju=i; if(ju>ncmax-1) ju=ncmax-1;
			for(j=1; j<=ju; ++j) t[i][j]=t[i][j-1]+(t[i][j-1]-t[i-1][j-1])/
				(h[i-j]*h[i-j]/(h[i]*h[i])-1.);

			df=t[i][ju];
			err1=err;
			j=i-1; if(j>ncmax-1) j=ncmax-1;
			e1=fabs(df-t[i-1][j]);
			err=fabs(df-t[i][ju-1]); if(e1<err) err=e1;
			e1=fabs(df)*reps; if(e1<aeps) e1=aeps;
			if(err<e1) return df;

			if(i>4 && err>2.*err1) {
/*	Roundoff error dominates */
				*ier=29;
				return t[i-1][j];
			}
		}
	}
	*ier=28;
	return df;
}

File: projects/antia_numerical_methods@Beliavsky/f77/xmp4/drvt.f
!     PROGRAM FOR DIFFERENTIATION USING H --> 0 EXTRAPOLATION

      PROGRAM DIFF
!      IMPLICIT REAL*8(A-H,O-Z)
      EXTERNAL F

!     EXAMPLE 5.2: DERIVATIVES OF EXP(X) 

51    FORMAT('   IER =',I4,I6,'TH DERIVATIVE AT X =',1PE14.6,
     1  3X,1PE14.6/ '   EXACT VALUE =',E14.6,5X,'INITIAL SPACING ='
     2  ,E14.6) 
 
      REPS=1.E-5
      AEPS=1.E-6

100   PRINT *,'TYPE  ID=ORDER OF DERIVATIVE,    (QUITS WHEN ID.LE.0)'
      READ *,ID
      IF(ID.LE.0) STOP
      PRINT *,'A=REQUIRED POINT,   HH=INITIAL SPACING'
      READ *,A,HH
      DF=DRVT(A,ID,HH,REPS,AEPS,F,IER)
      WRITE(6,51) IER,ID,A,DF,EXP(A),HH
      GO TO 100
      END
 
!     --------------------------------------------------
 
!	To calculate derivative of a function
!
!	A : (input) Value of x at which the derivative needs to be calculated
!	ID : (input) Order of derivative required, ID may be 1,2,3 or 4
!	HH0 : (input) Initial value of step length to be used.
!	REPS : (input) Required relative accuracy
!	AEPS : (input) Required absolute accuracy.
!		Calculations will continue till estimated error is less
!		than MAX(AEPS, REPS*ABS(DRVT))
!	F : (input) The name of function routine to calculate the given function
!		FUNCTION F(X) must be supplied by the user.
!	IER : (output) Error parameter, IER=0 for successful execution
!		IER=28 implies required accuracy is not achieved
!		IER=29 implies roundoff error appears to be dominating
!		IER=208 implies ID<1 or ID>4, no calculations are done.
!
!	DRVT is the calculated derivative at x=A
!		
!	FUNCTION F(X) must be supplied by the user
!
!	Required routines : F

      FUNCTION DRVT(A,ID,HH0,REPS,AEPS,F,IER)
!      IMPLICIT REAL*8(A-H,O-Z)
      PARAMETER(NMAX=15,NCMAX=6,HFAC=1.5)
      DIMENSION H(NMAX),T(NMAX,NCMAX)

!	Set spacing to the specified initial value
      HH=HH0
      ERR=0.0
      IER=0

      DO 2000 I=1,NMAX
        H(I)=HH

        IF(ID.EQ.1) THEN
          T(I,1)=(F(A+HH)-F(A-HH))/(2.*HH)
        ELSE IF(ID.EQ.2) THEN
          T(I,1)=(F(A-HH)-2.*F(A)+F(A+HH))/(HH*HH)
        ELSE IF(ID.EQ.3) THEN
          T(I,1)=(F(A+2.*HH)-F(A-2.*HH)-2.*(F(A+HH)-F(A-HH)))/(2.*HH**3)
        ELSE IF(ID.EQ.4) THEN
          T(I,1)=(F(A+2*HH)+F(A-2*HH)-4*(F(A+HH)+F(A-HH))+6*F(A))/HH**4
        ELSE
          IER=208
          RETURN
        ENDIF

!	Reduce spacing by a factor of HFAC
        HH=HH/HFAC
        IF(I.GT.1) THEN
!	Calculate at most NCMAX columns of the T-table
          JU=MIN(I,NCMAX)
          DO 1200 J=2,JU
           T(I,J)=T(I,J-1)+(T(I,J-1)-T(I-1,J-1))/((H(I-J+1)/H(I))**2-1.)
1200      CONTINUE

          DRVT=T(I,JU)
          ERR1=ERR
          ERR=MIN(ABS(DRVT-T(I,JU-1)),ABS(DRVT-T(I-1,MIN(I-1,NCMAX))))
          IF(ERR.LT.MAX(ABS(DRVT)*REPS,AEPS)) RETURN

          IF(I.GT.5.AND.ERR.GT.2.*ERR1) THEN
!	Roundoff error dominates
            IER=29
            DRVT=T(I-1,MIN(I-1,NCMAX))
            RETURN
          ENDIF
        ENDIF

2000  CONTINUE
      IER=28
      END
 
!     ------------------------------------------------------
 
      FUNCTION F(X)
!      IMPLICIT REAL*8(A-H,O-Z)

!     THE GIVEN FUNCTION

      F=EXP(X)
      END
